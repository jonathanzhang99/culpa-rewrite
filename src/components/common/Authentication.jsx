import PropTypes from "prop-types";
import React, { useContext, useReducer } from "react";
import { Route, Redirect } from "react-router-dom";

export const AuthContext = React.createContext(null);

function authReducer(state, action) {
  switch (action.type) {
    case "LOGIN_START":
      return {
        ...state,
        isLoading: true,
      };
    case "LOGIN_ERROR":
      return {
        ...state,
        isError: true,
        isLoading: false,
      };

    case "LOGIN_SUCCESS":
      localStorage.setItem("authToken", JSON.stringify(action.payload));
      return {
        ...state,
        isError: false,
        isLoading: false,
        isLoggedIn: true,
        authToken: action.payload,
      };

    case "LOGIN_FAILURE":
      return {
        ...state,
        isError: false,
        isLoading: false,
        isLoggedIn: false,
      };
    case "LOGOUT_SUCCESS":
      localStorage.removeItem("authToken");
      return {
        ...state,
        isError: false,
        isLoading: false,
        isLoggedIn: false,
        authToken: undefined,
      };
    default:
      throw new Error();
  }
}

const propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.arrayOf(PropTypes.element),
  ]).isRequired,
};

/**
 * AuthProvider createa React Context with values such that children elements
 * in the tree can access user auth token and manipulate them
 * (e.g. login, logout)
 *
 */
export function AuthProvider({ children }) {
  const existingToken = JSON.parse(localStorage.getItem("authToken"));

  const [state, dispatch] = useReducer(authReducer, {
    isError: false,
    isLoading: false,
    isLoggedIn: !!existingToken,
    authToken: existingToken,
  });

  // `login` either returns an error or null if a token was succesfully
  // retrieved. The error will always be a string message
  const login = async ({ username, password }) => {
    dispatch({ type: "LOGIN_START" });
    const response = await fetch("/api/auth/login", {
      method: "POST",
      body: JSON.stringify({
        username,
        password,
      }),
      headers: {
        "Content-Type": "application/json",
      },
    });
    try {
      const result = await response.json();

      // Require that the backend sends either a token or an error
      result.error
        ? dispatch({ type: "LOGIN_FAILURE" })
        : dispatch({ type: "LOGIN_SUCCESS", payload: result.token });
      return result;
    } catch (err) {
      dispatch({ type: "LOGIN_ERROR" });
      return err;
    }
  };

  // `logout` simply removes the token from the client. It does NOT
  // expire the token.
  const logout = async () => {
    if (state.authToken) {
      dispatch({ type: "LOGOUT_SUCCESS" });
    }
  };

  return (
    <AuthContext.Provider
      value={{
        ...state,
        login,
        logout,
      }}
    >
      {children}
    </AuthContext.Provider>
  );
}

AuthProvider.propTypes = propTypes;

/**
 * ProtectedRoute allows access to the url only if a user is logged in.
 *
 */
export function ProtectedRoute(props) {
  const { isLoggedIn } = useContext(AuthContext);
  // eslint-disable-next-line react/jsx-props-no-spreading
  return isLoggedIn ? <Route {...props} /> : <Redirect to="/" />;
}
