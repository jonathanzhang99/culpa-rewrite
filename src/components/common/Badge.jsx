import PropTypes from "prop-types";
import React from "react";
import { Image } from "semantic-ui-react";

import bronzeNugget from "icons/bronze-nugget.svg";
import goldNugget from "icons/gold-nugget.svg";
import silverNugget from "icons/silver-nugget.svg";

const propTypesBadge = {
  badgeId: PropTypes.number.isRequired,
};

export default function Badge({ badgeId }) {
  switch (badgeId) {
    case 1: {
      return <Image avatar spaced="left" src={goldNugget} />;
    }
    case 2: {
      return <Image avatar spaced="left" src={silverNugget} />;
    }
    case 3: {
      return <Image avatar spaced="left" src={bronzeNugget} />;
    }
    default: {
      return null;
    }
  }
}

Badge.propTypes = propTypesBadge;
