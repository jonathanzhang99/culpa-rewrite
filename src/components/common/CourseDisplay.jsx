import PropTypes from "prop-types";
import React from "react";
import { Link } from "react-router-dom";
import { Header } from "semantic-ui-react";

const propTypesCourseDisplayName = {
  as: PropTypes.string,
  name: PropTypes.string.isRequired,
  code: PropTypes.string,
};

const propTypesCourseDisplayLink = {
  as: PropTypes.string,
  courseId: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  code: PropTypes.string,
};

const defaultProps = {
  as: "p",
  code: "",
};

export function CourseDisplayName({ as, code, name }) {
  return (
    <Header as={as}>
      {code} {name}
    </Header>
  );
}

export function CourseDisplayLink({ as, courseId, code, name }) {
  return (
    <Link
      to={{
        pathname: `/courses/${courseId}`,
      }}
    >
      <CourseDisplayName as={as} code={code} name={name} />
    </Link>
  );
}

CourseDisplayName.propTypes = propTypesCourseDisplayName;
CourseDisplayName.defaultProps = defaultProps;

CourseDisplayLink.propTypes = propTypesCourseDisplayLink;
CourseDisplayLink.defaultProps = defaultProps;
