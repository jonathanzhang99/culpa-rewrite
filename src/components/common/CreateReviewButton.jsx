/* eslint-disable react/jsx-props-no-spreading */
import PropTypes from "prop-types";
import React from "react";
import { Link } from "react-router-dom";
import { Button } from "semantic-ui-react";

const propTypes = {
  courseId: PropTypes.number,
  professorId: PropTypes.number,
};

const defaultProps = {
  courseId: undefined,
  professorId: undefined,
};

export default function CreateReviewButton({ courseId, professorId, ...rest }) {
  return (
    <Link
      to={{
        pathname: "/review/submit",
        state: { courseId, professorId },
      }}
    >
      <Button {...rest} />
    </Link>
  );
}

CreateReviewButton.propTypes = propTypes;
CreateReviewButton.defaultProps = defaultProps;
