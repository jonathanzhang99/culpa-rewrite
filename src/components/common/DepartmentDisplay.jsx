import PropTypes from "prop-types";
import React from "react";
import { Link } from "react-router-dom";
import { Header } from "semantic-ui-react";

const propTypesDepartmentDisplayName = {
  as: PropTypes.oneOf(["header", "text"]),
  departmentName: PropTypes.string.isRequired,
};

const propTypesDepartmentDisplayLink = {
  as: PropTypes.oneOf(["header", "text"]),
  departmentId: PropTypes.number.isRequired,
  departmentName: PropTypes.string.isRequired,
};

const defaultProps = {
  as: "text",
};

export function DepartmentDisplayName({ as, departmentName }) {
  const DepartmentName = as === "header" ? Header : "span";
  return <DepartmentName> {departmentName} </DepartmentName>;
}

export function DepartmentDisplayLink({ as, departmentId, departmentName }) {
  return (
    <Link to={`/department/${departmentId}`}>
      <DepartmentDisplayName as={as} departmentName={departmentName} />
    </Link>
  );
}

DepartmentDisplayName.propTypes = propTypesDepartmentDisplayName;
DepartmentDisplayName.defaultProps = defaultProps;

DepartmentDisplayLink.propTypes = propTypesDepartmentDisplayLink;
DepartmentDisplayLink.defaultProps = defaultProps;
