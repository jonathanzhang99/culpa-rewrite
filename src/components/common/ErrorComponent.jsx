import React from "react";

export default function ErrorComponent() {
  return <div>error occurred</div>;
}
