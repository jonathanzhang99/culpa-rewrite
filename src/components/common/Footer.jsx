import React from "react";
import { Link } from "react-router-dom";
import { Grid, Icon, Image } from "semantic-ui-react";

import culpaIcon from "icons/culpa.svg";

export default function Footer() {
  return (
    <Grid className="footer" columns={2} verticalAlign="middle">
      <Grid.Column textAlign="center">
        <Image centered src={culpaIcon} />
        <Icon color="blue" name="copyright outline" size="small" />
        <span style={{ color: "#82B7E8" }}>1997-2020</span>
      </Grid.Column>
      <Grid.Column textAlign="left">
        <Grid.Row>
          <Link className="footer-link" to="/">
            About
          </Link>
        </Grid.Row>
        <Grid.Row>
          <Link className="footer-link" to="/join-the-team">
            Join The Team
          </Link>
        </Grid.Row>
        <Grid.Row>
          <Icon color="blue" name="mail" size="small" />
          <span className="footer-link">admin@culpa.info</span>
        </Grid.Row>
      </Grid.Column>
    </Grid>
  );
}
