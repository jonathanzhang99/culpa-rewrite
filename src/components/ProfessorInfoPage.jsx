import PropTypes from "prop-types";
import React from "react";
import { useParams } from "react-router-dom";
import { Grid, Header } from "semantic-ui-react";

import { CourseDisplayLink } from "components/common/CourseDisplay";
import CreateReviewButton from "components/common/CreateReviewButton";
import ErrorComponent from "components/common/ErrorComponent";
import LoadingComponent from "components/common/LoadingComponent";
import { ProfessorDisplayName } from "components/common/ProfessorDisplay";
import useDataFetch from "components/common/useDataFetch";
import ReviewCard from "components/reviews/ReviewCard";
import ReviewSection from "components/reviews/ReviewSection";

const propTypesProfessorCourses = {
  courses: PropTypes.arrayOf(
    PropTypes.shape({
      courseId: PropTypes.number.isRequired,
      courseName: PropTypes.string.isRequired,
      courseCallNumber: PropTypes.string.isRequired,
    })
  ).isRequired,
};

function ProfessorCourseList({ courses }) {
  return (
    <div className="add-margin">
      <Header className="no-margin">Courses: </Header>
      {courses.map(({ courseId, courseName, courseCallNumber }, index) => (
        <span key={courseId}>
          <CourseDisplayLink
            courseCallNumber={courseCallNumber}
            courseId={courseId}
            courseName={courseName}
          />
          {index !== courses.length - 1 ? ", " : ""}
        </span>
      ))}
    </div>
  );
}

const propTypesProfessorSummary = {
  badges: PropTypes.arrayOf(PropTypes.number).isRequired,
  courses: PropTypes.arrayOf(
    PropTypes.shape({
      courseId: PropTypes.number.isRequired,
      courseName: PropTypes.string.isRequired,
      courseCallNumber: PropTypes.string.isRequired,
    })
  ).isRequired,
  firstName: PropTypes.string.isRequired,
  lastName: PropTypes.string.isRequired,
  professorId: PropTypes.number.isRequired,
};

export function ProfessorSummary({
  badges,
  courses,
  firstName,
  lastName,
  professorId,
}) {
  return (
    /* use div to move to a new line */
    <>
      <div>
        <ProfessorDisplayName
          as="header"
          badges={badges}
          firstName={firstName}
          lastName={lastName}
          size="huge"
        />
      </div>
      <div>
        <ProfessorCourseList courses={courses} />
      </div>
      <div>
        <CreateReviewButton
          relaxed
          color="orange"
          professorId={professorId}
          subject={`${firstName} ${lastName}`}
        />
      </div>
    </>
  );
}

const reviewPropType = PropTypes.shape({
  reviewType: PropTypes.string.isRequired,
  reviewHeader: PropTypes.shape({
    courseId: PropTypes.number.isRequired,
    courseName: PropTypes.string.isRequired,
    courseCallNumber: PropTypes.string.isRequired,
  }).isRequired,
  votes: PropTypes.shape({
    initUpvoteCount: PropTypes.number.isRequired,
    initDownvoteCount: PropTypes.number.isRequired,
    initFunnyCount: PropTypes.number.isRequired,
    upvoteClicked: PropTypes.bool.isRequired,
    downvoteClicked: PropTypes.bool.isRequired,
    funnyClicked: PropTypes.bool.isRequired,
  }).isRequired,
  workload: PropTypes.string,
  submissionDate: PropTypes.string.isRequired,
  reviewId: PropTypes.number.isRequired,
  deprecated: PropTypes.bool,
  content: PropTypes.string,
});

const propTypesReview = {
  review: reviewPropType.isRequired,
};

function ProfessorReviewCard({ review }) {
  return (
    <ReviewCard
      content={review.content}
      deprecated={review.deprecated}
      reviewHeader={review.reviewHeader}
      reviewId={review.reviewId}
      reviewType={review.reviewType}
      submissionDate={review.submissionDate}
      votes={review.votes}
      workload={review.workload}
    />
  );
}

function SingleProfessorReviewHighlight({ review }) {
  return (
    <Grid columns={2}>
      <Grid.Column key="most_agreed_review_highlight">
        <Header>Most Agreed Review</Header>
        <ProfessorReviewCard review={review} />
      </Grid.Column>
    </Grid>
  );
}

const propTypesProfessorReviewHighlight = {
  professorReviewHighlight: PropTypes.arrayOf(reviewPropType).isRequired,
};

function DoubleProfessorReviewHighlight({ professorReviewHighlight }) {
  return (
    <Grid columns={2}>
      <Grid.Column key="most_positive_review_highlight">
        <Header>Most Positive Review</Header>
        <ProfessorReviewCard review={professorReviewHighlight[0]} />
      </Grid.Column>
      <Grid.Column key="most_negative_review_highlight">
        <Header>Most Negative Review</Header>
        <ProfessorReviewCard review={professorReviewHighlight[1]} />
      </Grid.Column>
    </Grid>
  );
}

function ProfessorReviewHighlight({ professorReviewHighlight }) {
  /* 
    professorReviewHighlight can have lengths 0-2:
      2: professorReviewHighlight[0] is the most positive review and
          professorReviewHighlight[1] the most negative review
      1: Either most positive and negative reviews are the same review
          or there is only one review for the course
      0: There are no reviews for the course
  */

  if (professorReviewHighlight.length === 1) {
    return (
      <SingleProfessorReviewHighlight review={professorReviewHighlight[0]} />
    );
  }
  if (professorReviewHighlight.length === 2) {
    return (
      <DoubleProfessorReviewHighlight
        professorReviewHighlight={professorReviewHighlight}
      />
    );
  }
  return <></>;
}

export default function ProfessorInfoPage() {
  const { professorId } = useParams();
  const professorDataFetched = useDataFetch(`/api/professor/${professorId}`, {
    professorSummary: {
      firstName: "",
      lastName: "",
      badges: [],
      courses: [],
    },
    professorReviewHighlight: [],
  });

  const reviewDataFetched = useDataFetch(
    `/api/review/get/professor/${professorId}`,
    {
      reviews: [],
    }
  );

  const {
    professorSummary,
    professorReviewHighlight,
  } = professorDataFetched.data;
  const isProfessorLoading = professorDataFetched.isLoading;
  const isProfessorError = professorDataFetched.isError;

  const { reviews } = reviewDataFetched.data;
  const isReviewLoading = reviewDataFetched.isLoading;
  const isReviewError = reviewDataFetched.isError;

  if (isProfessorLoading || isProfessorError) {
    return isProfessorLoading ? <LoadingComponent /> : <ErrorComponent />;
  }

  if (isReviewLoading || isReviewError) {
    return isReviewLoading ? <LoadingComponent /> : <ErrorComponent />;
  }

  return (
    <>
      <ProfessorSummary
        badges={professorSummary.badges}
        courses={professorSummary.courses}
        firstName={professorSummary.firstName}
        lastName={professorSummary.lastName}
        professorId={Number(professorId)}
      />
      <ProfessorReviewHighlight
        professorReviewHighlight={professorReviewHighlight}
      />
      <ReviewSection
        associatedEntities={professorSummary.courses}
        id={Number(professorId)}
        initReviews={reviews}
        pageType="professor"
      />
    </>
  );
}

ProfessorCourseList.propTypes = propTypesProfessorCourses;

ProfessorSummary.propTypes = propTypesProfessorSummary;

ProfessorReviewCard.propTypes = propTypesReview;

SingleProfessorReviewHighlight.propTypes = propTypesReview;

DoubleProfessorReviewHighlight.propTypes = propTypesProfessorReviewHighlight;

ProfessorReviewHighlight.propTypes = propTypesProfessorReviewHighlight;
