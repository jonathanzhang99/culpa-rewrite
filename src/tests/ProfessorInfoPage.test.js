import { render } from "@testing-library/react";
import React from "react";
import { MemoryRouter, Route } from "react-router-dom";

import { AuthContext } from "components/common/Authentication";
import ProfessorInfoPage from "components/ProfessorInfoPage";

describe("ProfessorInfoPage Component", () => {
  const loginSuccess = jest.fn(() => {
    Promise.resolve();
  });
  const TEST_PROFESSOR_ID_PARAM = 1;
  const testCases = [
    {
      name: "renders no info",
      firstName: "",
      lastName: "",
      courses: [{}],
    },
    {
      name: "renders professor info",
      firstName: "Nakul",
      lastName: "Verma",
      courses: [
        {
          courseProfessorId: 1,
          courseName: "Machine Learning",
          courseCallNumber: "COMS 4771",
        },
        {
          courseProfessorId: 2,
          courseName: "Advanced Machine Learning",
          courseCallNumber: "COMS 4774",
        },
      ],
    },
  ];
  testCases.forEach(({ name, firstName, lastName, courses }) => {
    test(name, () => {
      const snapshot = render(
        <MemoryRouter initialEntries={[TEST_PROFESSOR_ID_PARAM]}>
          <Route path=":professorId">
            <AuthContext.Provider value={{ login: loginSuccess }}>
              <ProfessorInfoPage
                courses={courses}
                firstName={firstName}
                lastName={lastName}
              />
            </AuthContext.Provider>
          </Route>
        </MemoryRouter>
      );
      expect(snapshot).toMatchSnapshot();
    });
  });
});
