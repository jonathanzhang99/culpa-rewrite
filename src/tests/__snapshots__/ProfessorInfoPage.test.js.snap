// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfessorSummary Component renders professor info 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="ui header display-header"
      >
        Nakul
         
        Verma
      </div>
      <div
        class="ui container"
      >
        <span>
          Courses: 
        </span>
        <span>
          <a
            href="/course/1"
          >
            <span
              class="display-header"
            >
              <span
                class="sc-AxjAm ekCnfx"
              >
                COMS 4771
              </span>
               
              Machine Learning
            </span>
          </a>
          , 
        </span>
        <span>
          <a
            href="/course/2"
          >
            <span
              class="display-header"
            >
              <span
                class="sc-AxjAm ekCnfx"
              >
                COMS 4774
              </span>
               
              Advanced Machine Learning
            </span>
          </a>
          
        </span>
      </div>
      <a
        href="/review"
      >
        <button
          class="ui yellow button"
        >
          WRITE A REVIEW FOR 
          Nakul
           
          Verma
        </button>
      </a>
    </div>
  </body>,
  "container": <div>
    <div
      class="ui header display-header"
    >
      Nakul
       
      Verma
    </div>
    <div
      class="ui container"
    >
      <span>
        Courses: 
      </span>
      <span>
        <a
          href="/course/1"
        >
          <span
            class="display-header"
          >
            <span
              class="sc-AxjAm ekCnfx"
            >
              COMS 4771
            </span>
             
            Machine Learning
          </span>
        </a>
        , 
      </span>
      <span>
        <a
          href="/course/2"
        >
          <span
            class="display-header"
          >
            <span
              class="sc-AxjAm ekCnfx"
            >
              COMS 4774
            </span>
             
            Advanced Machine Learning
          </span>
        </a>
        
      </span>
    </div>
    <a
      href="/review"
    >
      <button
        class="ui yellow button"
      >
        WRITE A REVIEW FOR 
        Nakul
         
        Verma
      </button>
    </a>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`ProfessorSummary Component renders professor no courses 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="ui header display-header"
      >
        Nakul
         
        Verma
      </div>
      <div
        class="ui container"
      >
        <span>
          Courses: 
        </span>
      </div>
      <a
        href="/review"
      >
        <button
          class="ui yellow button"
        >
          WRITE A REVIEW FOR 
          Nakul
           
          Verma
        </button>
      </a>
    </div>
  </body>,
  "container": <div>
    <div
      class="ui header display-header"
    >
      Nakul
       
      Verma
    </div>
    <div
      class="ui container"
    >
      <span>
        Courses: 
      </span>
    </div>
    <a
      href="/review"
    >
      <button
        class="ui yellow button"
      >
        WRITE A REVIEW FOR 
        Nakul
         
        Verma
      </button>
    </a>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
